File: testdata/precision-suite-kotlin/clean_parse_id.kt
// Clean: the user-controlled value is coerced to an integer with String.toInt()
// before use, which throws on any non-numeric input and so strips every shell /
// SQL / path metacharacter. The value reaching exec is a validated Int rendered
// back to a string — no injection is possible. A correct scanner emits nothing;
// any TAINT-002 finding here is a false positive.
package com.example

import javax.servlet.http.HttpServletRequest

class JobRunner {
    fun run(request: HttpServletRequest) {
        val raw = request.getParameter("jobId")
        val jobId = raw.toInt()
        Runtime.getRuntime().exec("run-job --id " + jobId)
    }
}

File: testdata/precision-suite-rust/clean_parse_id.rs
// Clean: the user-controlled value is coerced to a typed integer with
// str::parse::<i64>() before being interpolated into SQL. Numeric coercion
// removes every injection metacharacter, so no SQLi is possible even though a
// format! builds the query. A correct scanner emits nothing; a TAINT-001
// finding here is a false positive.
use std::env;

async fn lookup_by_id(pool: &sqlx::PgPool) {
    let raw = env::var("USER_ID").unwrap_or_default();
    let id: i64 = raw.parse().unwrap_or(0);
    // `id` is an i64 — it cannot carry a SQL payload.
    let sql = format!("SELECT * FROM users WHERE id = {}", id);
    let rows = sqlx::query(&sql).fetch_all(pool).await;
    let _ = rows;
}
